File: leetcode-04/solutions/cpp/0023.cpp
class Solution {
 public:
  ListNode* mergeKLists(vector<ListNode*>& lists) {
    ListNode dummy(0);
    ListNode* curr = &dummy;
    priority_queue<ListNode*, vector<ListNode*>, compareListNode> pq;

    for (auto list : lists)
      if (list) pq.push(list);

    while (!pq.empty()) {
      curr->next = pq.top();
      pq.pop();
      curr = curr->next;
      if (curr->next) pq.push(curr->next);
    }

    return dummy.next;
  }

 private:
  struct compareListNode {
    bool operator()(const ListNode* l1, const ListNode* l2) {
      return l1->val > l2->val;
    }
  };
};

File: leetcode-04/solutions/python3/0023.py
from heapq import heappush, heappop


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        dummy = ListNode(0)
        curr = dummy
        heap = []

        for i in range(len(lists)):
            if lists[i]:
                heappush(heap, (lists[i].val, i, lists[i]))

        while heap:
            node = heappop(heap)
            curr.next = node[2]
            curr = curr.next
            if curr.next:
                heappush(heap, (curr.next.val, node[1], curr.next))

        return dummy.next
